verification_tokens/decorators.py

```diff
@@ -0,0 +1,89 @@
+from functools import wraps
+from typing import Callable, Optional
+
+from django.core.exceptions import PermissionDenied
+
+from common.utils import context
+from verification_tokens.models import VerificationToken
+
+
+def user_from_auth_verification_token(
+    get_token: Callable[[dict], Optional[str]], use_only_when_first_denied: bool = False
+):
+    """Sets the related auth token user to the context
+    (in place of the logged in user; ``context.user``),
+    if the wrapped function throws a permission denied error and then retries
+    the wrapped function.
+
+    This decorator is implemented to be used with the ``@login_required`` decorator
+    in the GraphQL endpoint's handler.
+
+    Example
+    =======
+    In this example, the decorator first executes the ``@login_required``,
+    and then if it raises the ``PermissionDenied`` error,
+    the ``"auth_token"`` is used from the input variables to find
+    an authorization verification token that links to an user that
+    the token represents
+    >>> # doctest: +SKIP
+    ... # When the login_required raises a PermissionDenied exception,
+    ... # use the auth_token from the input variables
+    ... # to populate the context.user with the token related user
+    ... # and then try again.
+    ... @user_from_auth_verification_token_when_denied(
+    ...     get_token=lambda variables: variables.get("auth_token", None)
+    ... )
+    ... @login_required
+    ... def mutate_and_get_payload(cls, root, info, **kwargs):
+    ...     pass
+
+    The input in this example would be
+    >>> # doctest: +SKIP
+    ... {
+    ...     "input": {
+    ...         "authToken": auth_verification_token.key,
+    ...     }
+    ... }
+
+    Args:
+        get_token ((dict) -> str | None): a (lambda) function that
```

Comment:
![image](https://github.com/City-of-Helsinki/kukkuu/assets/77663720/349f4a66-a094-4f22-a245-2761b35c2cfa)

There's a non-breaking space in `str | None` just before `None`.